File: foliage-rpc/src/main/java/io/foliage/netty/rpc/core/AbortPolicyWithReport.java
package io.foliage.netty.rpc.core;

import lombok.extern.slf4j.Slf4j;

import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.ThreadPoolExecutor;

@Slf4j
public class AbortPolicyWithReport extends ThreadPoolExecutor.AbortPolicy {

    private final String threadName;

    public AbortPolicyWithReport(String threadName) {
        this.threadName = threadName;
    }

    public void rejectedExecution(Runnable r, ThreadPoolExecutor e) {
        String msg = String.format("RpcServer["
                        + " Thread Name: %s, Pool Size: %d (active: %d, core: %d, max: %d, largest: %d), Task: %d (completed: %d),"
                        + " Executor status:(isShutdown:%s, isTerminated:%s, isTerminating:%s)]",
                threadName, e.getPoolSize(), e.getActiveCount(), e.getCorePoolSize(),
                e.getMaximumPoolSize(), e.getLargestPoolSize(), e.getTaskCount(),
                e.getCompletedTaskCount(), e.isShutdown(), e.isTerminated(), e.isTerminating());
        log.error(msg);

        throw new RejectedExecutionException(msg);
    }
}

File: foliage-rpc/src/main/kotlin/io/foliage/netty/rpc/core/AbortPolicyWithReport.kt
package io.foliage.netty.rpc.core

import org.lavenderx.foliage.nettyrpc.utils.loggerFor
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.ThreadPoolExecutor

class AbortPolicyWithReport(val threadName: String) : ThreadPoolExecutor.AbortPolicy() {

    private val logger = loggerFor<AbortPolicyWithReport>()

    override fun rejectedExecution(r: Runnable, e: ThreadPoolExecutor) {
        val msg = String.format("RpcServer["
                + " Thread Name: %s, Pool Size: %d (active: %d, core: %d, max: %d, largest: %d), Task: %d (completed: %d),"
                + " Executor status: (isShutdown: %s, isTerminated: %s, isTerminating: %s)]",
                threadName, e.poolSize, e.activeCount, e.corePoolSize, e.maximumPoolSize, e.largestPoolSize,
                e.taskCount, e.completedTaskCount, e.isShutdown, e.isTerminated, e.isTerminating)
        logger.error(msg)

        throw RejectedExecutionException(msg)
    }
}